[JrmCkh] iP

JrmCkh opened the pull request:
# Duke 😎 
#### A personal assistant chatbot that helps you to keep track of various things
> ######  _Why remember things, when Duke can do it for you? 🤔_

Duke is:
- simple user interface
- text-based
- easy to learn
- ~~FAST~~ ***SUPER FAST*** to use

Duke features:
- [ ] Managing tasks
- [x] Managing deadlines
- [x] Reminders (coming soon)

All you have to do is:
1. download it form [here](https://github.com/nus-cs2103-AY2223S2/ip/pull/321)
2. double-click it
3. add your tasks
4. let it manage for you!

What's more? It is totally ***FREE*** 😄

Commit:
Add Gradle support

JrmCkh committed:
Add Greet feature

JrmCkh committed:
Add Greet, Echo, and Exit feature

JrmCkh committed:
Add adding and listing of tasks feature

JrmCkh committed:
Add Task and Feedback classes

JrmCkh committed:
Add mark task as done feature

JrmCkh committed:
Add Todo, Deadline, and Event classes

JrmCkh committed:
Add tracking of Todo, Deadline, and Event tasks

JrmCkh committed:
Add DukeException class to handle errors

JrmCkh committed:
Add TaskList class for storing of data

JrmCkh committed:
Add tasks deletion feature

JrmCkh committed:
Fix wrong feedback from deletion of task

JrmCkh committed:
Implement semi-automate testing of Duke

JrmCkh committed:
Add Storage class and tasks text file to save tasks

JrmCkh committed:
Implement saving tasks to storage feature

JrmCkh committed:
Add date and time feature

JrmCkh committed:
Update Feedback and Task class

JrmCkh committed:
Merge branch 'branch-Level-7'

JrmCkh committed:
Merge branch 'branch-Level-8'

JrmCkh committed:
Add Command, Parser, and Ui classes, and remove Feedback class

JrmCkh committed:
Modify code for Command, Parser and Ui class

JrmCkh committed:
Add classes into packages

JrmCkh committed:
Merge remote-tracking branch 'origin/add-gradle-support'

* origin/add-gradle-support:
  Add Gradle support

JrmCkh committed:
Fix the bug of wait for user input after exit program

JrmCkh committed:
Modify main class name to Duke in build.gradle

JrmCkh committed:
Fix bug of showing D instead of E in toString

JrmCkh committed:
Add JUnit test to test code behavior

JrmCkh committed:
Add manifest file

JrmCkh committed:
Add JavaDoc for Duke and command classes

JrmCkh committed:
Add JavaDoc

JrmCkh committed:
Improve coding standard

JrmCkh committed:
Add find task by keyword feature

JrmCkh committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc
  Add JavaDoc for Duke and command classes

JrmCkh committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Improve coding standard

JrmCkh committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find task by keyword feature

JrmCkh committed:
Add configuration files for checkstyle

JrmCkh committed:
Update to meet checkstyle rule

JrmCkh committed:
Merge branch 'branch-A-CheckStyle'

* branch-A-CheckStyle:
  Update to meet checkstyle rule
  Add configuration files for checkstyle

JrmCkh committed:
Add a GUI to Duke using JavaFX

JrmCkh committed:
Modify to use JavaFX GUI

JrmCkh committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Modify to use JavaFX GUI
  Add a GUI to Duke using JavaFX

zhengsterz commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,58 @@
+package duke;
+
+import duke.commands.Command;
+import duke.parser.Parser;
+import duke.storage.Storage;
+import duke.task.TaskList;
+import duke.ui.Ui;
```
Perhaps the import statements can be separated via a space for better clarity?

zhengsterz commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,191 @@
+     * @param input the user input
+     * @return the command based on the input
+     */
+    public Command parseCommand(String input) {
+        CommandCode commandCode = getCommand(input);
+        switch (commandCode) {
+        case TODO:
+            return prepareTodo(input.substring(4).trim());
+        case DEADLINE:
+            return prepareDeadline(input.substring(8).trim());
+        case EVENT:
+            return prepareEvent(input.substring(5).trim());
+        case DELETE:
+            return prepareDelete(input.substring(6).trim());
+        case MARK:
+            return prepareMark(input.substring(4).trim());
+        case UNMARK:
+            return prepareUnmark(input.substring(6).trim());
+        case FIND:
+            return prepareFind(input.substring(4).trim());
+        case LIST:
+            return new ListCommand();
+        case HELP:
+            return new HelpCommand();
+        case EXIT:
+            return new ExitCommand();
+        default:
+            return new InvalidCommand();
+        }
+    }
```
I like how clean your parser is when handling this portion as I am only using if else statements

zhengsterz commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @author Chia Jeremy
+ */
+public class Event extends Task {
```
Perhaps consider having the suffix -Task added to classes that inherit from Task for clarity? E.g EventTask

zhengsterz commented on text-ui-test/EXPECTED.TXT:
```diff
@@ -1,7 +1,65 @@
+    DISPLAY COMMANDS:  help
+    EXIT PROGRAM:      bye
+    ____________________________________________________________
+
```
Adding a legend for possible commands is a great idea

zhengsterz commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,67 @@
+     * @param description the description of the task
+     * @param dateTime    the deadline to finish the task
+     */
+    public AddCommand(String description, LocalDateTime dateTime) {
```
Cool how you extended the DateTime function to all Tasks

zhengsterz reviewed (commented):
Very neat and efficient code!

BigheartSmallbrain commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,67 @@
+        storage.add(this.task);
+        ui.setResponse("Got it. I've added this task:\n"
+                + this.task + "\n"
+                + "Now you have " + tasks.getSize() + " tasks in the list.");
```
There's some repetition in printing the message for number of tasks in the list. I think maybe it would be better to store such messages and just call it.

BigheartSmallbrain commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,191 @@
+     * @param input the user input
+     * @return the command based on the input
+     */
+    public Command parseCommand(String input) {
```
It seems that there's some form of inconsistency in this block of code. I think it might look neater if prepare() can change into like new TodoCommand() like the ListCommand(). What do you think?

BigheartSmallbrain commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,140 @@
+    public void mark(int index) {
+        List<String> tasks = getSavedTasksAsList();
+        StringBuilder sb = new StringBuilder(tasks.get(index));
+        sb.setCharAt(4, 'X');
```
Is this a magic number? I think it could have been assigned with varible name such as startingIndex.

BigheartSmallbrain reviewed (commented):
The codes are well-written and named overall. I'm impressed with the level of consistency in the variable and method namings that explain what the variable or method do without going through the block of codes.

JrmCkh committed:
Remove magic numbers and improve code consistency

JrmCkh committed:
Update JAR file v0.2 to include GUI

JrmCkh committed:
Add asserts

JrmCkh committed:
Improve code quality

JrmCkh committed:
Merge pull request #2 from JrmCkh/branch-A-CodeQuality

Improve code quality

JrmCkh committed:
Merge branch 'master' into branch-A-Assertions

JrmCkh committed:
Merge pull request #1 from JrmCkh/branch-A-Assertions

Add asserts

JrmCkh committed:
Update indentation

JrmCkh committed:
Add GitHub Actions to set up Continuous Integration

Continuous Integration allows finding and addressing bugs quicker,
improve software quality, and reduce the time it takes to validate
and release updates.

Let's, add a gradle.yml file to make use of GitHub Actions

JrmCkh committed:
Use GitHub Actions to set up Continuous Integration

Continuous Integration allows for finding and addressing
bugs quicker, improve software quality, and reduce the time
it takes to validate and release new updates

Let's add a gradle.yml file to make use of GitHub Actions feature

WesleyBLDC commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+    public Duke(String filePath) {
+        assert pathNotEmpty(filePath);
+        this.storage = new Storage(filePath);
+        this.tasks = new TaskList(storage.load());
```
perhaps it wld be more clear using this.storage

JrmCkh committed:
Add help command to provide in-app guidance to user

JrmCkh committed:
Merge branch 'branch-C-Help'

* branch-C-Help:
  Add help command to provide in-app guidance to user

JrmCkh committed:
Add the feature of automatically sort by task, then by date

JrmCkh committed:
Add SLAP principles

JrmCkh committed:
Merge branch 'branch-C-Sort'

* branch-C-Sort:
  Add SLAP principles
  Add the feature of automatically sort by task, then by date

JrmCkh committed:
Change height of text box to show full text

JrmCkh committed:
Add better gui

JrmCkh committed:
Merge branch 'branch-A-BetterGui'

* branch-A-BetterGui:
  Add better gui
  Change height of text box to show full text

JrmCkh committed:
Add more testing

JrmCkh committed:
Merge branch 'branch-A-MoreTesting'

* branch-A-MoreTesting:
  Add more testing

JrmCkh committed:
Add ReadMe and Ui image

JrmCkh committed:
Update README

JrmCkh committed:
Add testing for Parser class

JrmCkh committed:
Update duke response for list command

JrmCkh committed:
Add auto create file path and file for saving tasks
